dnf module doesn't honor state: present / state: installed

jameslivulpi opened the pull request:
##### SUMMARY
Fixes #76463 

When using `state: present` or `state: installed` the check on if the package is installed always returns `False`

```python3

if installed.filter(**package_spec):
  return True
else:
  return False
```

The dict unpacking of `package_spec` caused this. This fix removes the unpacking.

##### ISSUE TYPE

- Bugfix Pull Request


##### COMPONENT NAME
dnf

##### ADDITIONAL INFORMATION
Details in issue mentioned, but before change running the following play:

```
- hosts: all
  tasks:
    - dnf:
        name: java-11-openjdk
        state: installed
```

java-11-openjdk is already installed before the play:
```
[root@centos-s-1vcpu-1gb-nyc1-01 ~]# rpm -qa | grep java-11-openjdk
java-11-openjdk-11.0.13.0.8-1.el8_4.x86_64
java-11-openjdk-headless-11.0.13.0.8-1.el8_4.x86_64
```

Updates the packages which shouldn't happen:

```
    "msg": "",
    "rc": 0,
    "results": [
        "Installed: java-11-openjdk-headless-1:11.0.13.0.8-3.el8_5.x86_64",
        "Installed: java-11-openjdk-1:11.0.13.0.8-3.el8_5.x86_64",
        "Removed: java-11-openjdk-1:11.0.13.0.8-1.el8_4.x86_64",
        "Removed: java-11-openjdk-headless-1:11.0.13.0.8-1.el8_4.x86_64"
    ]
```


After fix:

```
[root@centos-s-1vcpu-1gb-nyc1-01 ~]# rpm -qa | grep jdk
java-11-openjdk-11.0.13.0.8-1.el8_4.x86_64
java-11-openjdk-headless-11.0.13.0.8-1.el8_4.x86_64
```
Run play and returns:

```
        }
    },
    "msg": "Nothing to do",
    "rc": 0,
    "results": []
}

```

packages still at same level:
```
[root@centos-s-1vcpu-1gb-nyc1-01 ~]# rpm -qa | grep jdk
java-11-openjdk-11.0.13.0.8-1.el8_4.x86_64
java-11-openjdk-headless-11.0.13.0.8-1.el8_4.x86_64
```

jameslivulpi committed:
fix dict unpacking

add changelog

reworked present/installed handling and added integration test

Add hardness detection of installed packages

use get_best_query for checking on installed

jameslivulpi commented:
azp /run

jameslivulpi commented:
/azp run

jameslivulpi commented:
/rebuild

samdoran commented:
The tests are failing due to the changes made in this PR. The packages needed to run the `dnf` test are not being installed since the `dnf` module was changed.

jameslivulpi commented:
> The tests are failing due to the changes made in this PR. The packages needed to run the `dnf` test are not being installed since the `dnf` module was changed.

yes thanks, figured out the exact test and reason and pushing shortly

jameslivulpi commented:
ok apologies for all the commits, still getting used to the test flow, think I got it now :-)

This takes care of current issue mentioned, I do see some areas in this code I may open another pr at a later point.

s-hertel commented on test/integration/targets/dnf/tasks/dnf.yml:
```diff
@@ -581,6 +581,41 @@
+    name: httpd
+    state: removed
+
+- name: Verify installed/present state is not changed
```
These assertions are passing on devel for me, so I don't think the bug is getting tested. The issue https://github.com/ansible/ansible/issues/76463 also seems to be running into the problem in check mode, so we should make sure check mode gives the same 'changed' result as non-check mode.

1kaw replied:
https://github.com/ansible/ansible/issues/76463#issuecomment-988340340

s-hertel replied:
Didn't mean we should replace these, just that we should *also* be testing check mode.

jameslivulpi commented:
I can see the issue clearly now.

Here's Fedora35:

```
[root@fedora-s-1vcpu-1gb-nyc1-01 ~]# dnf install java-11-openjdk
Last metadata expiration check: 0:00:10 ago on Tue 07 Dec 2021 04:25:41 PM UTC.
Package java-11-openjdk-1:11.0.12.0.7-4.fc35.x86_64 is already installed.
Dependencies resolved.
Nothing to do.
Complete!
```

On CentOS the behavior is different..

```
[root@centos-s-1vcpu-1gb-nyc1-01 ~]#  dnf install java-11-openjdk
Last metadata expiration check: 0:01:42 ago on Tue 07 Dec 2021 04:25:57 PM UTC.
Package java-11-openjdk-1:11.0.13.0.8-1.el8_4.x86_64 is already installed.
Dependencies resolved.
==========================================================================================================================================================================
 Package                                          Architecture                   Version                                          Repository                         Size
==========================================================================================================================================================================
Upgrading:
 java-11-openjdk                                  x86_64                         1:11.0.13.0.8-3.el8_5                            appstream                         266 k
 java-11-openjdk-headless                         x86_64                         1:11.0.13.0.8-3.el8_5                            appstream                          39 M

Transaction Summary
==========================================================================================================================================================================
Upgrade  2 Packages

Total size: 40 M
Is this ok [y/N]: 
```


This is due to `/etc/dnf/dnf.conf` defaults:

Fedora35:
```
best=False
```

CentOS 8:
```
best=True
```

Changing CentOS node to use `best=False` and I get what I would expect.
```
[root@centos-s-1vcpu-1gb-nyc1-01 ~]#  dnf install java-11-openjdk
Last metadata expiration check: 0:01:19 ago on Tue 07 Dec 2021 04:41:11 PM UTC.
Package java-11-openjdk-1:11.0.13.0.8-1.el8_4.x86_64 is already installed.
Dependencies resolved.
Nothing to do.
Complete!
```

Discussion here:
https://github.com/rpm-software-management/dnf/pull/1484

Should we change this in ansible? I guess if `state == installed || present` we could force the config to nobest but not sure if that's a correct way forward.

jameslivulpi commented:
Adding more to above after further testing. I think the bug is here:

``` python3
        # Set best
        if self.nobest:
            conf.best = 0
```

and should be:
```python3
       # Set best
        if not self.nobest:   
            conf.best = 0
```
By default self.nobest is False so ,so we should set dnf config best to False.

Ran some unit tests locally, so far so good:

Downgraded java-11-openjdk on CentOS 8 and Fedora 35 nodes:

**Test on my branch**

Play:
```
- hosts: all
  tasks:
    - dnf:
        name:
          - java-11-openjdk
        state: installed
```

 java-11-openjdk did not get upgrade ✔️ 

```
(venv) ➜  ansible git:(dnfpresentfix) ✗ ansible-playbook dnf.yml  -i hosts -u root      
[WARNING]: You are running the development version of Ansible. You should only run Ansible from "devel" if you are modifying the Ansible engine, or trying out features under development. This is a rapidly changing source of code and can become unstable at any point.

PLAY [all] ***********************************************************************************************************************************************************************************************************************************************************************************

TASK [Gathering Facts] ***********************************************************************************************************************************************************************************************************************************************************************
ok: [159.223.127.187]
ok: [167.172.139.237]

TASK [dnf] ***********************************************************************************************************************************************************************************************************************************************************************************
ok: [159.223.127.187]
ok: [167.172.139.237]

PLAY RECAP ***********************************************************************************************************************************************************************************************************************************************************************************
159.223.127.187            : ok=2    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
167.172.139.237            : ok=2    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
```



**Test on devel**
Same play of above..

```
(venv) ➜  ansible git:(devel) ✗ ansible-playbook dnf.yml  -i hosts -u root     
[WARNING]: You are running the development version of Ansible. You should only run Ansible from "devel" if you are modifying the Ansible engine, or trying out features
under development. This is a rapidly changing source of code and can become unstable at any point.

PLAY [all] ***************************************************************************************************************************************************************

TASK [Gathering Facts] ***************************************************************************************************************************************************
ok: [159.223.127.187]
ok: [167.172.139.237]

TASK [dnf] ***************************************************************************************************************************************************************
ok: [167.172.139.237]
changed: [159.223.127.187]

PLAY RECAP ***************************************************************************************************************************************************************
159.223.127.187            : ok=2    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
167.172.139.237            : ok=2    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
```


Would like to see what others think?

relrod commented:
> By default self.nobest is False so ,so we should set dnf config best to False.

It's inverted, though. self.**no**best (emphasis on the **no**). So when nobest is true, best should be false. When nobest is false, best should be true.

See also https://github.com/ansible/ansible/pull/74224 which probably needs to be dusted off if someone wants to take it over. I went down the rabbit hole a fair bit there, and added some comments and a bunch of test cases. Some parts of it touch the `nobest` logic.

jameslivulpi commented:
Again sorry for all commits to this.. but I try to be descriptive in each update so its less confusing :-)

Summary of the fix:

- updated `_packagename_dict` method for better handling of getting package details if the package is currently installed on the system and if user is calling with state: present or state: installed.
- Updated `_whatprovides` method and added an argument `available` which by default (available=True) maintains prior behavior such as getting any available package that's given.  Using available=False  allows for checking on a package a user has given in the form of "java-11-openjdk" for only the **currently installed packages** and gets the NEVRA format if there is an installed packages that provides the given one. This addresses an issue I saw where in previous code the NEVRA could have incorrect values mentioned in prior [comment](https://github.com/ansible/ansible/pull/76467#issuecomment-990346210)
- Add output to stdout when the package is indeed already installed as this is good for user and follows "Always return useful data, even when there is no change." Example below.
- Added test cases and few relating to the java-11-openjdk package specifically.


play:
```
   tasks:      
     - dnf:   
         name:
           - java-11-openjdk
           - httpd      
           - vim
         state: installed
```

output when all packages are already installed:

```
    "msg": "Nothing to do",
    "rc": 0,
    "results": [
        "Package vim is already installed.",
        "Package java-11-openjdk is already installed.",
        "Package httpd is already installed."
```

Output when java-11-openjdk and vim are already installed but not httpd:

```
    "msg": "",
    "rc": 0,
    "results": [
        "Package java-11-openjdk is already installed.",
        "Package vim is already installed.",
        "Installed: centos-logos-httpd-85.8-2.el8.noarch",
        "Installed: mod_http2-1.15.7-3.module_el8.4.0+778+c970deab.x86_64",
        "Installed: apr-1.6.3-12.el8.x86_64",
        "Installed: apr-util-1.6.1-6.el8.x86_64",
        "Installed: apr-util-bdb-1.6.1-6.el8.x86_64",
        "Installed: httpd-2.4.37-43.module_el8.5.0+1022+b541f3b1.x86_64",
        "Installed: httpd-filesystem-2.4.37-43.module_el8.5.0+1022+b541f3b1.noarch",
        "Installed: mailcap-2.1.48-3.el8.noarch",
        "Installed: apr-util-openssl-1.6.1-6.el8.x86_64",
        "Installed: httpd-tools-2.4.37-43.module_el8.5.0+1022+b541f3b1.x86_64"
    ]
}
```

1kaw commented:
Why needs_revision?
Because @s-hertel requested changes?

marekvesely-direct commented:
Is there any progress? This issue is very crucial for us to see if there are any changes on managed nodes. @s-hertel

LudekPalenik commented:
Hello. I would like to ask if the issue is being addressed in any way. It has been open for almost 2 years now, and there have been no changes so far. It's somewhat inconvenient to constantly verify changes that essentially do not exist. Thank you.

jameslivulpi commented:
rebased, ci passed.

webknjaz commented:
/azp run

marekvesely-direct commented:
@webknjaz When this request will be merged?

pstrzelczak commented:
@jameslivulpi this branch has conflicts. @s-hertel could you review the changes once the conflicts get resolved? This damages our experience with Ansble package module significantly.

mkrizek commented:
Superseded by https://github.com/ansible/ansible/pull/82725
